nxontology_ml/gpt_tagger/_gpt_tagger.py

```diff
@@ -0,0 +1,119 @@
+import logging
+from collections import Counter
+from collections.abc import Callable, Iterable
+from copy import deepcopy
+
+from nxontology.node import NodeInfo, NodeT
+
+from nxontology_ml.gpt_tagger._cache import _Cache
+from nxontology_ml.gpt_tagger._chat_completion_middleware import (
+    _ChatCompletionMiddleware,
+)
+from nxontology_ml.gpt_tagger._models import LabelledNode, TaskConfig
+from nxontology_ml.gpt_tagger._tiktoken_batcher import _TiktokenBatcher
+from nxontology_ml.gpt_tagger._utils import (
+    counter_or_empty,
+    efo_id_from_yaml,
+    node_efo_id,
+    node_to_str_fn,
+    parse_model_output,
+)
+
+
+class GptTagger:
+    """
+    Main entry point for node tagging using GPT models.
+    Fetches labels for given nodes.
+
+    Implementation details:
+        The tagger relies on the following classes:
+            - :class:`_ChatCompletionMiddleware`: Handle communications with the Open AI API.
+            - :class:`_Cache`: Allows caching of the labels that have already been fetched. By default, the caches are
+                on disk in the `./.cache` directory.
+            - :class:`_TiktokenBatcher`: This utility class allows to count the amount of token of the records to be
+                labelled. This token count is used for record batching.
+
+    """
+
+    def __init__(
+        self,
+        chat_completion_middleware: _ChatCompletionMiddleware,
+        cache: _Cache,
+        batcher: _TiktokenBatcher,
+        node_to_str_fn: Callable[[NodeInfo[NodeT]], str],
+        counter: Counter[str],
+    ):
+        """
+        Intended to get constructed using cls.from_config(config)
+        """
+        self._chat_completion_middleware = chat_completion_middleware
+        self._cache = cache
+        self._batcher = batcher
+        self._node_to_str_fn = node_to_str_fn
+        self._counter = counter
+
+    def fetch_labels(self, nodes: Iterable[NodeInfo[NodeT]]) -> Iterable[LabelledNode]:
+        """
+        Fetch the labels (using GPT) for the provided nodes.
+        """
+        buffer: list[str] | None
+        for node in nodes:
+            node_str = self._node_to_str_fn(node)
```

Comment:
So the caching works on an individual node level. If a future query includes a node with the same id, label, and description there will be a cache hit. With the current prompt this makes the most sense (as compared to caching prompt output, which would be susceptible to misses due to variable batching).